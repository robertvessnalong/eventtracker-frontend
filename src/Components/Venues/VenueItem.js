import { Card, Button, Col, Row } from 'react-bootstrap';

const VenueItem = ({ venue }) => {
  return (
    <Card className='mt-4 mb-4 p-0'>
      <Card.Header as='h6'>
        <Row>
          <Col className='d-flex align-items-center' xs={10}>
            <small>{venue.name}</small>
          </Col>
        </Row>
      </Card.Header>
      <Card.Body>
        <Row>
          <Col xs={12} md={6} className='text-start'>
            <address>
              {venue.address}
              <br />
              {venue.extended_address}
            </address>
          </Col>
          <Col
            xs={12}
            md={6}
            className='d-flex align-items-center justify-content-end'
          >
            <a href={venue.url} rel='noreferrer' target='_blank'>
              <Button className='blue-btn' variant='primary'>
                Buy Tickets
              </Button>
            </a>
          </Col>
        </Row>
      </Card.Body>
    </Card>
  );
};

export default VenueItem;
