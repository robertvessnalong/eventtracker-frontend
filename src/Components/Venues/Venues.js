import { useState, useEffect } from 'react';
import {
  Container,
  Row,
  Col,
  Spinner,
  FloatingLabel,
  Form,
  Button,
  Alert,
} from 'react-bootstrap';
import EventFinderApi from '../../API/api';
import VenueItem from './VenueItem';
import uuid from 'react-uuid';

const Venues = () => {
  const formInitalState = {
    search: '',
  };
  const [page, setPage] = useState(1);
  const [formData, setFormData] = useState(formInitalState);
  const [venues, setVenues] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(false);
  const [errorMsg, setErrorMsg] = useState([]);
  const [hasVenues, setHasVenues] = useState(false);

  useEffect(() => {
    const getVenues = async () => {
      const res = await EventFinderApi.getVenues({ page });
      setVenues(res);
      setLoading(false);
    };

    getVenues();
  }, [page]);

  if (loading) {
    return (
      <Row>
        <Col style={{ textAlign: 'center' }}>
          <Spinner animation='border' role='status'>
            <span className='visually-hidden'>Loading...</span>
          </Spinner>
        </Col>
      </Row>
    );
  }

  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((data) => ({
      ...data,
      [name]: value,
    }));
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    const res = await EventFinderApi.getVenues(formData);
    if (res.error) {
      setError(true);
      setErrorMsg(res.error);
    } else {
      if (venues.length > 0) {
        setHasVenues(false);
        setFormData(formInitalState);
        setVenues(res);
      } else {
        setHasVenues(true);
      }
    }
  };

  const handlePagination = (e) => {
    if (e.target.classList.contains('btn')) {
      const attr = e.target.getAttribute('aria-label');
      if (attr === 'Previous' && page !== 1) {
        setPage((prevState) => prevState - 1);
      } else if (attr === 'Next' && page >= 1) {
        setPage((prevState) => prevState + 1);
      }
    }
  };

  return (
    <div className='Venues'>
      <header>
        <h2>Venues</h2>
        <span>Find A Venue Near You</span>
        <hr></hr>
      </header>
      <Container>
        <div className='Search'>
          <Form onSubmit={handleSubmit} className='mt-4'>
            {error && <Alert variant='danger'>{errorMsg}</Alert>}
            <Row className='g-2'>
              <Col md>
                <FloatingLabel
                  controlId='floatingInputGrid'
                  label='Search Venues...'
                >
                  <Form.Control
                    type='text'
                    name='search'
                    value={formData.search}
                    onChange={handleChange}
                  />
                </FloatingLabel>
              </Col>
              <Col md={3}>
                <Button
                  type='submit'
                  className='h-100 w-100 blue-btn'
                  variant='primary'
                >
                  Find Venue
                </Button>{' '}
              </Col>
            </Row>
          </Form>
        </div>
      </Container>
      <Container className='Event-Items'>
        {hasVenues && (
          <Alert className='mt-4' variant='warning'>
            Sorry, no events found! Try Again
          </Alert>
        )}
        <Row xs={1} xl={2}>
          {venues.map((venue) => (
            <Col key={uuid()}>
              <VenueItem venue={venue} />
            </Col>
          ))}
        </Row>
        <Row className='mt-5'>
          <Col>
            <div aria-label='pagination'>
              <ul
                className='pagination justify-content-center'
                key={uuid()}
                onClick={handlePagination}
              >
                <Button
                  disabled={page === 1 ? true : false}
                  className='page-link'
                  href='#'
                  aria-label='Previous'
                  key={uuid()}
                >
                  <span className='page-arrow' aria-hidden='true'>
                    &laquo;
                  </span>
                </Button>
                <Button
                  className='page-link'
                  href='#'
                  aria-label='Next'
                  key={uuid()}
                >
                  <span className='page-arrow' aria-hidden='true'>
                    &raquo;
                  </span>
                </Button>
              </ul>
            </div>
          </Col>
        </Row>
      </Container>
    </div>
  );
};

export default Venues;
