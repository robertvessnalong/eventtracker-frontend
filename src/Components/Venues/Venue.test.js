import {
  render,
  screen,
  fireEvent,
  waitForElementToBeRemoved,
  waitFor,
} from '@testing-library/react';
import '@testing-library/jest-dom';
import Venues from './Venues';
import TestUser from '../../Testing/TestUser';
import UserContext from '../../Context/UserContext';
import { BrowserRouter } from 'react-router-dom';

describe('Venue Component', () => {
  // Smoke Test
  test('renders without crashing', async () => {
    render(
      <BrowserRouter>
        <UserContext.Provider value={{ user: TestUser }}>
          <Venues />
        </UserContext.Provider>
      </BrowserRouter>
    );
    await waitForElementToBeRemoved(screen.getByText('Loading...'));
    const label = screen.getByText(/Find A Venue Near You/i);
    expect(label).toBeInTheDocument();
  });

  // Snapshot
  test('matches snapshot', async () => {
    const { asFragment } = render(
      <BrowserRouter>
        <UserContext.Provider value={{ user: TestUser }}>
          <Venues />
        </UserContext.Provider>
      </BrowserRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('fireSubmit', async () => {
    const { getByText, getByRole } = render(
      <BrowserRouter>
        <UserContext.Provider value={{ user: TestUser }}>
          <Venues />
        </UserContext.Provider>
      </BrowserRouter>
    );
    await waitForElementToBeRemoved(screen.getByText('Loading...'));
    await waitFor(() => {
      fireEvent.change(getByRole('textbox', { name: /search/i }), {
        query: '',
      });
      fireEvent.submit(getByText('Find Venue'));
    });
    const alertMsg = screen.getByText(/McGuirk Arena/i);
    expect(alertMsg).toBeInTheDocument();
  });

  test('updated snapshot', async () => {
    const { asFragment } = render(
      <BrowserRouter>
        <UserContext.Provider value={{ user: TestUser }}>
          <Venues />
        </UserContext.Provider>
      </BrowserRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
