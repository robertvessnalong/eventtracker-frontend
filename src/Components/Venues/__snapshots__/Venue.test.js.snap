// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Venue Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="row"
  >
    <div
      class="col"
      style="text-align: center;"
    >
      <div
        class="spinner-border"
        role="status"
      >
        <span
          class="visually-hidden"
        >
          Loading...
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Venue Component updated snapshot 1`] = `
<DocumentFragment>
  <div
    class="row"
  >
    <div
      class="col"
      style="text-align: center;"
    >
      <div
        class="spinner-border"
        role="status"
      >
        <span
          class="visually-hidden"
        >
          Loading...
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
